x-pack/auditbeat/processors/add_session_metadata/provider/ebpf_provider/ebpf_provider.go

```diff
@@ -0,0 +1,159 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package ebpf_provider
+
+import (
+	"context"
+
+	"github.com/mohae/deepcopy"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/add_session_metadata/processdb"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/add_session_metadata/provider"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/add_session_metadata/types"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/ebpfevents"
+)
+
+type prvdr struct {
+	ctx                context.Context
+	logger             *logp.Logger
+	db                 *processdb.DB
+}
+
+func NewProvider(ctx context.Context, logger *logp.Logger, db *processdb.DB) (provider.Provider, error) {
+	p := prvdr{
+		ctx:    ctx,
+		logger: logger,
+		db:     db,
+	}
+
+	l, err := ebpfevents.NewLoader()
+	if err != nil {
+		p.logger.Errorf("loading ebpf: %w", err)
+		return prvdr{}, err
+	}
+
+	events := make(chan ebpfevents.Event)
+	errors := make(chan error)
+
+	go l.EventLoop(p.ctx, events, errors)
+
+	go func(logger logp.Logger) {
+		for {
+			select {
+			case err := <-errors:
+				logger.Errorf("recv'd error: %w", err)
+				continue
+			case ev := <-events:
```

Comment:
It is a good idea to avoid non-trivial select statements if possible; a select with more than one non-default case incur significant scheduler costs. If the ebpfevents API can be altered to send on a single chan and have that hold either an error or an event with the addition of an error field, this would then resolve to a chan range loop.

Reply:
https://github.com/elastic/ebpfevents/pull/18

Reply:
hey @efd6 your comment here really got my attention because firstly I wasn't aware of it (I mean not in the magnitude I think you imply) and I know many codebases out there that utilise such a multi-case select for loop pattern. So that makes it really interesting. Do you have any link that showcase the details of it, so I can educate myself? :)

Reply:
This was from a comment on the gophers slack in performance. Comment by Josh on 2023-01-11.

Reply:
ebpfevents library has been updated to use a single channel, and I've updated this branch to use it